Add odoc dev tool

gridbugs opened the pull request:
The command `dune ocaml doc` will now lock, build, and run odoc as a dev tool if the dev tools feature is enabled.

gridbugs committed:
Generalize dev tool fetching

Remove references to ocamlformat from the fetch rules and create fetch
rules for all dev tools.

Signed-off-by: Stephen Sherratt <stephen@sherra.tt>

gridbugs committed:
Add odoc dev tool

The command `dune ocaml doc` will now lock, build, and run odoc as a
dev tool if the dev tools feature is enabled.

Signed-off-by: Stephen Sherratt <stephen@sherra.tt>

rgrinberg commented:
Doesn't odoc have the same limitation as lsp? I.e it needs to be compiled with the same version of OCaml as the project.

rgrinberg commented on src/dune_rules/odoc.ml:
```diff
@@ -245,14 +245,33 @@ let odoc_base_flags quiet build_dir =
 
+let odoc_dev_tool_lock_dir_exists () =
+  let path = Dune_pkg.Lock_dir.dev_tool_lock_dir_path Odoc in
+  Fs_memo.dir_exists (Path.source path |> Path.as_outside_build_dir_exn)
```
Don't think the `as_outside_build_dir_exn` is necessary. You can just use of the constructors for `Path.Outside_build_dir` directly.

gridbugs replied:
Good point

moyodiallo commented on test/blackbox-tests/test-cases/pkg/odoc/dev-tool-odoc-basic.t:
```diff
@@ -0,0 +1,35 @@
+executable to attempt to generate documentation. The fake executable
+doesn't actually do anything so the command fails but this
+demonstrates that it was run.
+  $ DUNE_CONFIG__PKG_BUILD_PROGRESS=enabled DUNE_CONFIG__LOCK_DEV_TOOL=enabled dune ocaml doc
```
```suggestion
  $ DUNE_CONFIG__LOCK_DEV_TOOL=enabled dune ocaml doc
```

gridbugs replied:
Good point, fixed.

gridbugs commented:
> Doesn't odoc have the same limitation as lsp? I.e it needs to be compiled with the same version of OCaml as the project.

You're totally right. I didn't realize this but that explains some of the weird errors I've run into with odoc recently:

E.g.
```
File "src/.climate.objs/byte/_unknown_", line 1, characters 0-0:
ERROR: File "climate__.cmt":
not an implementation
```

I'll borrow some of the code from my [ocamllsp dev-tool pr](https://github.com/ocaml/dune/pull/10915#pullrequestreview-2312160014) to make sure that the same compiler is used to build odoc as is used to build the project.

gridbugs commented:
I updated this so that it always uses the same version of the compiler to build odoc as it does to build the project and added some tests that it relocks the odoc dev tool to keep the compiler dependency in sync with the project.
